crates/sui-storage/src/indexes.rs

```diff
@@ -295,17 +295,26 @@ impl IndexStore {
         let coin_delete_keys = object_index_changes
             .deleted_owners
             .iter()
+            .dedup()
             .filter_map(|(owner, obj_id)| {
                 // If it's not in `input_coins`, then it's not a coin type. Skip.
-                let coin_type_tag = input_coins.get(obj_id)?
-                .coin_type_maybe().unwrap_or_else(|| {
+                let object = input_coins.get(obj_id)?;
+                let coin_type_tag = object.coin_type_maybe().unwrap_or_else(|| {
                     panic!(
                         "object_id: {:?} in input_coins is not a coin type, input_coins: {:?}, tx_digest: {:?}",
                         obj_id, input_coins, digest
                     )
                 });
-                address_coin_type_to_invalidate.insert((*owner, coin_type_tag.clone()));
-                addresses_to_invalidate.insert(*owner);
+                let map = balance_changes.entry(*owner).or_insert(HashMap::new());
+                let entry = map.entry(coin_type_tag.clone()).or_insert(TotalBalance {
+                    num_coins: 0,
+                    balance: 0
+                });
+                let existing = self.tables.coin_index.get(&(*owner, coin_type_tag.to_string(), *obj_id));
```

Comment:
would there be a data race here?

Reply:
There could be, although I am not 100% sure because it seems like input and output objects of one transaction cannot intersect with another txns for owned objects but not for shared objects. I added a lock now which prevents any possible race b/w concurrent coin index access for an address. This lock is held very briefly.